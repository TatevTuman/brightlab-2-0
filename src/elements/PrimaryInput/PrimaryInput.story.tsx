import React from 'react'
import { Meta, Story } from '@storybook/react/types-6-0'
import { storybook } from '@utils'
import PrimaryInput, { PrimaryInputProps } from './PrimaryInput'

interface PrimaryInputStoryTemplateProps extends PrimaryInputProps {}

export default {
  title: 'Elements/PrimaryInput',
  argTypes: {
    label: { control: 'text', name: 'Label' },
    placeholder: { control: 'text', name: 'Placeholder' },
    prefix: { control: 'text', name: 'Prefix' },
    suffix: { control: 'text', name: 'Suffix' },
    autoComplete: { control: storybook.controls.select(['on', 'off']), name: 'Autocomplete' },
    clearable: { control: 'boolean', name: 'Clearable' },
    loading: { control: 'boolean', name: 'Loading' },
    ...storybook.controls.interactive,
    onClear: { action: 'onClear' },
    name: storybook.args.disabled
  }
} as Meta

const PrimaryInputStoryTemplate: Story<PrimaryInputStoryTemplateProps> = args => {
  return <PrimaryInput {...args} />
}

PrimaryInputStoryTemplate.args = {
  placeholder: 'Type a text',
  ...storybook.actions.interactive,
  prefix: '',
  suffix: '',
  name: 'primary-input',
  autoComplete: 'on',
  clearable: true,
  loading: false
}

export const PrimaryInputStory = PrimaryInputStoryTemplate.bind({})
PrimaryInputStory.args = {
  ...PrimaryInputStoryTemplate.args
}
