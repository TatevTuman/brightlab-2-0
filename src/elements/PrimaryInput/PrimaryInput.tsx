import React, { memo, forwardRef, RefObject } from 'react'
import { Loader, Icon } from '@elements'
import { handleEvent } from '@utils'
import './PrimaryInput.scss'

export type PrimaryInputSuffixProp = JSX.Element | string | number
export interface PrimaryInputProps {
  name: string
  value: string
  type?: string
  label?: string
  defaultValue?: string
  placeholder?: string
  prefix?: PrimaryInputSuffixProp
  suffix?: PrimaryInputSuffixProp

  autoComplete?: 'on' | 'off'
  hideCursor?: boolean
  disabled?: boolean
  loading?: boolean
  error?: boolean
  clearable?: boolean
  required?: boolean
  role?: string
  ref?: RefObject<HTMLInputElement>

  onChange: (value: string) => void
  onFocus?: (e: React.FocusEvent<HTMLInputElement>) => void
  onBlur?: (e: React.FocusEvent<HTMLInputElement>) => void
  onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void
  onClear?: () => void
}

const PrimaryInput = forwardRef<HTMLInputElement, PrimaryInputProps>((props, ref) => {
  const {
    name,
    value,
    type,
    label,
    defaultValue,
    placeholder,
    prefix,
    suffix,
    autoComplete,
    hideCursor,
    disabled,
    error,
    loading,
    clearable,
    required,
    role,
    onChange,
    onFocus,
    onBlur,
    onKeyDown,
    onClear
  } = props

  /* If disabled no focus */
  const tabIndex = disabled ? -1 : 0

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    handleEvent(onChange, { value: e.currentTarget.value, disabled })

  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => handleEvent(onBlur, { value: e, disabled })
  const handleFocus = (e: React.FocusEvent<HTMLInputElement>) => handleEvent(onFocus, { value: e, disabled })
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => handleEvent(onKeyDown, { value: e, disabled })
  const handleClear = () => {
    if (!disabled) {
      onChange && onChange('')
    }
  }

  return (
    <label className={'primary-input'} data-disabled={disabled}>
      {label && (
        <label htmlFor={name} data-required={required} className=" fs-15 fw-600">
          {label}
        </label>
      )}
      <div className={'primary-input-inner'} data-error={!!error} data-testid={'inner'}>
        <div className={'primary-input-inner__prefix'}>{prefix}</div>
        <input
          id={name}
          type={type}
          name={name}
          ref={ref}
          value={value}
          defaultValue={defaultValue}
          placeholder={placeholder}
          autoComplete={autoComplete}
          onChange={handleChange}
          onKeyDown={handleKeyDown}
          onFocus={handleFocus}
          onBlur={handleBlur}
          data-nocursor={hideCursor}
          data-testid={'input'}
          role={role}
          tabIndex={tabIndex}
        />
        <div className={'primary-input-inner__suffix'}>
          {loading && <Loader type={'Oval'} width={16} height={16} />}
          {clearable && (
            <Icon
              name={'times'}
              className={'primary-input-inner__suffix-times'}
              data-testid={'primary-input-times'}
              onClick={onClear || handleClear}
            />
          )}
          {suffix}
        </div>
      </div>
    </label>
  )
})

PrimaryInput.displayName = 'PrimaryInput'
PrimaryInput.defaultProps = {
  type: 'text',
  autoComplete: 'on',
  clearable: true
}

export default memo(PrimaryInput)
