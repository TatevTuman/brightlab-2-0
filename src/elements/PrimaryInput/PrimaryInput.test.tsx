import React from 'react'
import { cleanup, render, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import PrimaryInput from './PrimaryInput'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('PrimaryInput', () => {
  it('renders correctly', async () => {
    const { container } = render(<PrimaryInput name={'input'} />)
    await waitFor(() => container)

    expect(container).toMatchSnapshot()
  })

  it('renders no label correctly', () => {
    const { container } = render(<PrimaryInput name={'input'} />)
    const label = container.querySelector('label')?.querySelector('label') || null

    expect(label).toBe(null)
  })

  it('renders label correctly', () => {
    const props = {
      name: 'input',
      label: 'input label'
    }

    const { getByText } = render(<PrimaryInput {...props} />)
    const label = getByText(props.label)
    expect(label).toBeInTheDocument()
  })

  it('renders required label correctly', () => {
    const props = {
      name: 'input',
      label: 'input label',
      required: true
    }
    const { container } = render(<PrimaryInput {...props} />)
    const label = container.querySelector('[data-required]') as HTMLInputElement

    expect(label).toHaveAttribute('data-required', 'true')
  })

  it('render error correctly', () => {
    const props = {
      name: 'input',
      error: true
    }
    const { getByTestId } = render(<PrimaryInput {...props} />)
    const inner = getByTestId('inner')
    expect(inner).toHaveAttribute('data-error', 'true')
  })

  it('renders disabled correctly', () => {
    const props = {
      name: 'input',
      label: 'input label',
      onFocus: jest.fn(),
      disabled: true
    }

    const { container, rerender } = render(<PrimaryInput {...props} />)
    const input = container.querySelector('[data-disabled]') as HTMLInputElement

    userEvent.tab()

    expect(props.onFocus).toHaveBeenCalledTimes(0)
    expect(input).toHaveAttribute('data-disabled', 'true')

    rerender(<PrimaryInput {...props} disabled={false} />)

    userEvent.tab()

    expect(props.onFocus).toHaveBeenCalledTimes(1)
    expect(input).toHaveAttribute('data-disabled', 'false')
  })

  it('handles onChange on label', () => {
    const props = {
      name: 'input',
      label: 'input label',
      onChange: jest.fn()
    }

    const { getByText, getByLabelText } = render(<PrimaryInput {...props} />)
    const label = getByText(props.label)
    const input = getByLabelText(props.label) as HTMLInputElement

    userEvent.click(label!)
    userEvent.type(label!, 'test')

    expect(props.onChange).toHaveBeenCalledTimes(4)
    expect(input.value).toBe('test')
  })

  it('handles onChange', () => {
    const props = {
      name: 'input',
      label: 'input label',
      onChange: jest.fn()
    }

    const { getByLabelText } = render(<PrimaryInput {...props} />)
    const input = getByLabelText(props.label) as HTMLInputElement

    userEvent.type(input!, 'test')

    expect(props.onChange).toHaveBeenCalledTimes(4)
    expect(input.value).toBe('test')
  })

  it('handles onFocus/onBlur', async () => {
    const props = {
      name: 'input',
      label: 'input label',
      onFocus: jest.fn(),
      onBlur: jest.fn()
    }
    render(<PrimaryInput {...props} />)

    /* Simulate tab */
    userEvent.tab()
    expect(props.onFocus).toHaveBeenCalledTimes(1)

    userEvent.tab()
    expect(props.onBlur).toHaveBeenCalledTimes(1)
  })
})
