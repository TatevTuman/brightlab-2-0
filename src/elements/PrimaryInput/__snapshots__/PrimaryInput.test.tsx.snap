// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryInput renders correctly 1`] = `
<div>
  <label
    class="primary-input"
  >
    <div
      class="primary-input-inner"
      data-error="false"
      data-testid="inner"
    >
      <div
        class="primary-input-inner__prefix"
      />
      <input
        autocomplete="on"
        data-testid="input"
        id="input"
        name="input"
        tabindex="0"
        type="text"
        value=""
      />
      <div
        class="primary-input-inner__suffix"
      >
        <i
          class="icon"
        >
          <test-file-stub
            classname="primary-input-inner__suffix-times"
            data-testid="primary-input-times"
            name="times"
          />
        </i>
      </div>
    </div>
  </label>
</div>
`;
